File: CS/GridExample/PersonsDataSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows.Input;
using DevExpress.Mvvm;

namespace GridExample {
    public class PersonsViewModel {
        ObservableCollection<Person> persons;
        
        public PersonsViewModel() {
            List<string> Names = new List<string> { "Alex", "Alice", "Tony", "Den", "Andrew", "John", "Donald", "Brian", "Effy", "Lisa", "Matthew" };
            persons = new ObservableCollection<Person>();
            for (int i = 0; i < 10; i++) {
                persons.Add(new Person(Names[i], "Last name " + i, 21 + i, i % 2 == 0, 170 + i, 75 + i));
            }
            persons[5].Age = 22;
            persons[8].Age = 50;
        }

        public ObservableCollection<Person> Persons {
            get {
                return persons;
            }
        }

        public ICommand ClearPersons {
            get {
                return new DelegateCommand<object>((p) => { Persons.Clear(); });
            }
        }
    }

    public class Person : INotifyPropertyChanged, IDataErrorInfo {
        const string lastNamePropertyName = "LastName";
        const string firstNamePropertyName = "FirstName";
        const string agePropertyName = "Age";
        const string heightPropertyName = "IsMarried";
        const string isMarriedPropertyName = "Height";
        const string weightPtropertyName = "Weight";

        string firstName;
        string lastName;
        int age;
        bool isMarried;
        int height;
        int weight;
        PersonPropertiesValidator Validator = new PersonPropertiesValidator();

        public Person() { }

        public Person(string firstName, string lastName, int age, bool isMarried, int height, int weight) {
            FirstName = firstName;
            LastName = lastName;
            Age = age;
            IsMarried = isMarried;
            Weight = weight;
            Height = height;
        }

        public string FirstName {
            get {
                return firstName;
            }
            set {
                Validator.IsNameValid(value, firstNamePropertyName);
                firstName = value;
                RaisePropertyChanged(firstNamePropertyName);
            }
        }

        public string LastName {
            get {
                return lastName;
            }
            set {
                Validator.IsNameValid(value, lastNamePropertyName);
                lastName = value;
                RaisePropertyChanged(lastNamePropertyName);
            }
        }

        public int Age {
            get {
                return age;
            }
            set {
                Validator.IsAgeValid(value, agePropertyName);
                age = value;
                RaisePropertyChanged(agePropertyName);
            }
        }

        public bool IsMarried {
            get {
                return isMarried;
            }
            set {
                isMarried = value;
                RaisePropertyChanged(isMarriedPropertyName);
            }
        }

        public int Height {
            get {
                return height;
            }
            set {
                height = value;
                RaisePropertyChanged(heightPropertyName);
            }
        }

        public int Weight {
            get {
                return weight;
            }
            set {
                weight = value;
                RaisePropertyChanged(weightPtropertyName);
            }
        }

        #region INotifyPropertyChanged members
        public event PropertyChangedEventHandler PropertyChanged;

        private void RaisePropertyChanged(string propertyName) {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
        #endregion

        #region IDataErrorInfo members
        public string Error {
            get {
                return Validator.DataError;
            }
        }

        public string this[string columnName] {
            get {
                if (Validator.DataErrors.ContainsKey(columnName))
                    return Validator.DataErrors[columnName];
                else
                    return null;
            }
        }
        #endregion
    }

    public class PersonPropertiesValidator {
        string dataError = "";
        Dictionary<string, string> dataErrors = new Dictionary<string, string>();

        public bool IsNameValid(string value, string propertyName) {
            if (string.IsNullOrEmpty(value)) {
                dataErrors[propertyName] = "Full name is required.";
                return false;
            } else {
                ClearPropertyErrors(propertyName);
                return true;
            }
        }

        public bool IsAgeValid(int value, string propertyName) {
            if (value <= 0) {
                dataErrors[propertyName] = "Age validation failed.";
                return false;
            } else {
                ClearPropertyErrors(propertyName);
                return true;
            }
        }

        public string DataError {
            get {
                return dataError;
            }
        }

        public Dictionary<string, string> DataErrors {
            get {
                return dataErrors;
            }
        }

        void ClearPropertyErrors(string propertyName) {
            if (dataErrors.ContainsKey(propertyName))
                dataErrors.Remove(propertyName);
        }
    }
}

File: VB/GridExample/PersonsDataSource.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Input
Imports DevExpress.Xpf.Core.Commands

Namespace GridExample

    Public Class PersonsViewModel

        Private personsField As ObservableCollection(Of Person)

        Public Sub New()
            Dim Names As List(Of String) = New List(Of String) From {"Alex", "Alice", "Tony", "Den", "Andrew", "John", "Donald", "Brian", "Effy", "Lisa", "Matthew"}
            personsField = New ObservableCollection(Of Person)()
            For i As Integer = 0 To 10 - 1
                personsField.Add(New Person(Names(i), "Last name " & i, 21 + i, i Mod 2 = 0, 170 + i, 75 + i))
            Next

            personsField(5).Age = 22
            personsField(8).Age = 50
        End Sub

        Public ReadOnly Property Persons As ObservableCollection(Of Person)
            Get
                Return personsField
            End Get
        End Property

        Public ReadOnly Property ClearPersons As ICommand
            Get
                Return New DelegateCommand(Of Object)(Sub(p) Persons.Clear())
            End Get
        End Property
    End Class

    Public Class Person
        Implements INotifyPropertyChanged, IDataErrorInfo

        Const lastNamePropertyName As String = "LastName"

        Const firstNamePropertyName As String = "FirstName"

        Const agePropertyName As String = "Age"

        Const heightPropertyName As String = "IsMarried"

        Const isMarriedPropertyName As String = "Height"

        Const weightPtropertyName As String = "Weight"

        Private firstNameField As String

        Private lastNameField As String

        Private ageField As Integer

        Private isMarriedField As Boolean

        Private heightField As Integer

        Private weightField As Integer

        Private Validator As PersonPropertiesValidator = New PersonPropertiesValidator()

        Public Sub New()
        End Sub

        Public Sub New(ByVal firstName As String, ByVal lastName As String, ByVal age As Integer, ByVal isMarried As Boolean, ByVal height As Integer, ByVal weight As Integer)
            Me.FirstName = firstName
            Me.LastName = lastName
            Me.Age = age
            Me.IsMarried = isMarried
            Me.Weight = weight
            Me.Height = height
        End Sub

        Public Property FirstName As String
            Get
                Return firstNameField
            End Get

            Set(ByVal value As String)
                Validator.IsNameValid(value, firstNamePropertyName)
                firstNameField = value
                RaisePropertyChanged(firstNamePropertyName)
            End Set
        End Property

        Public Property LastName As String
            Get
                Return lastNameField
            End Get

            Set(ByVal value As String)
                Validator.IsNameValid(value, lastNamePropertyName)
                lastNameField = value
                RaisePropertyChanged(lastNamePropertyName)
            End Set
        End Property

        Public Property Age As Integer
            Get
                Return ageField
            End Get

            Set(ByVal value As Integer)
                Validator.IsAgeValid(value, agePropertyName)
                ageField = value
                RaisePropertyChanged(agePropertyName)
            End Set
        End Property

        Public Property IsMarried As Boolean
            Get
                Return isMarriedField
            End Get

            Set(ByVal value As Boolean)
                isMarriedField = value
                RaisePropertyChanged(isMarriedPropertyName)
            End Set
        End Property

        Public Property Height As Integer
            Get
                Return heightField
            End Get

            Set(ByVal value As Integer)
                heightField = value
                RaisePropertyChanged(heightPropertyName)
            End Set
        End Property

        Public Property Weight As Integer
            Get
                Return weightField
            End Get

            Set(ByVal value As Integer)
                weightField = value
                RaisePropertyChanged(weightPtropertyName)
            End Set
        End Property

'#Region "INotifyPropertyChanged members"
        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Private Sub RaisePropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

'#End Region
'#Region "IDataErrorInfo members"
        Public ReadOnly Property [Error] As String Implements IDataErrorInfo.[Error]
            Get
                Return Validator.DataError
            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal columnName As String) As String Implements IDataErrorInfo.Item
            Get
                If Validator.DataErrors.ContainsKey(columnName) Then
                    Return Validator.DataErrors(columnName)
                Else
                    Return Nothing
                End If
            End Get
        End Property
'#End Region
    End Class

    Public Class PersonPropertiesValidator

        Private dataErrorField As String = ""

        Private dataErrorsField As Dictionary(Of String, String) = New Dictionary(Of String, String)()

        Public Function IsNameValid(ByVal value As String, ByVal propertyName As String) As Boolean
            If String.IsNullOrEmpty(value) Then
                dataErrorsField(propertyName) = "Full name is required."
                Return False
            Else
                ClearPropertyErrors(propertyName)
                Return True
            End If
        End Function

        Public Function IsAgeValid(ByVal value As Integer, ByVal propertyName As String) As Boolean
            If value <= 0 Then
                dataErrorsField(propertyName) = "Age validation failed."
                Return False
            Else
                ClearPropertyErrors(propertyName)
                Return True
            End If
        End Function

        Public ReadOnly Property DataError As String
            Get
                Return dataErrorField
            End Get
        End Property

        Public ReadOnly Property DataErrors As Dictionary(Of String, String)
            Get
                Return dataErrorsField
            End Get
        End Property

        Private Sub ClearPropertyErrors(ByVal propertyName As String)
            If dataErrorsField.ContainsKey(propertyName) Then dataErrorsField.Remove(propertyName)
        End Sub
    End Class
End Namespace
